### PR TITLE
feat: basic Safe transaction decoder

### PR DESCRIPTION
**Pull request by tkporter**
### Description

Basic functionality to verify multisig txs:
- Works with multiple txs from multiple chains (though atm, just 1 tx per chain in the output)
- Loudly complains if there's anything unknown
- Known scenarios are:
  - Multicalls
  - ICA txs
  - router enrollment
  - default ISM setting

Example usage:
```
yarn tsx ./scripts/safes/parse-txs.ts -e mainnet3 --chains ethereum mantapacific bsc arbitrum optimism base blast linea mode fraxtal taiko sei zircuit --txHashes 0x7f9344cc0b3fa68ece139fb9f0d2cf854c983c343a7680c6ab7b4bfb3924c188 0x1e8e00b7375085a3e3579659815611d1e8240aa43c8826d05347a3066603330d 0x39324715341f856466ca2764468bc31c3f057c8191c99ddad284714730f9337b 0x54eeae0cf702f6a9a3929712e196bb51e0d34778f5948ae8562e4456bb649c14 0x425d8fd82fb390a91f170c32a3bb8839dc9182ad9634111d477af85320d04612 0xbd29fe0f7387ac06bea2049eb26fb9c7b62ed4e412f84137de2a94fe917adca3 0xf6f05f757c22af8a306545d20238ecf3326e93d12aaeecf64d1527daa20c1ec6 0xd9bc3cf2f7ab01560fc99e57c25113d07283245554c4c133f8aa501fb0c21147 0xc6ed50e5ccdd001022f953ae31d7f0bbf88af78bc3dae593b62ad11725cfc900 0xb703be9d342b126f7596ce5431f5ead2192a95c92df9f3c1b76f5242ce8bba4f 0x662b126f552cdc5fcdc9fdad382e28c249a511786e0056dca2bb75ac452878a3 0x246e805047ece2522d98c6841cebf7ad1ce83b4503533a77d73e1b4324ea3ce8 0x412f787db9abdc9d4d4cf863b514790b7e57d5c1e2d9a963018c8b9a3a523fb9
```

Definitely messy and intended to be the basis of better tooling in the future.

### CODE REVIEW

File: typescript/infra/scripts/safes/ethereum-safe-tx-oct-30.json
Comment (paulbalaji) [inline typescript/infra/scripts/safes/ethereum-safe-tx-oct-30.json:1]: surely you can read this via the api directly

File: typescript/infra/src/tx/transaction-reader.ts [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+          value: tx.value,
+        },
+      );
+
+    const args = formatFunctionFragmentArgs(
+      decoded.args,
+      decoded.functionFragment,
+    );
+    let prettyArgs = args;
+    if (decoded.functionFragment.name === 'enrollRemoteRouters') {
```
Comment (Mo-Hussain) [inline typescript/infra/src/tx/transaction-reader.ts:100]: nit: can we import something that would avoid comparing to a literal?

**Comment by Mo-Hussain** [comment]
Can you share some outputs please!!

**Comment by paulbalaji** [comment]
> Can you share some outputs please!!

example outputs:
- https://discord.com/channels/935678348330434570/1301327529520664576/1301910909442068491
- https://discord.com/channels/935678348330434570/1301327529520664576/1301940555219079228

File: typescript/infra/scripts/safes/ethereum-safe-tx-oct-30.json
Comment (tkporter) [reply typescript/infra/scripts/safes/ethereum-safe-tx-oct-30.json:1]: ha yeah it does do this, I'll remove this

File: typescript/infra/src/tx/transaction-reader.ts (hunk 1 above)
Comment (tkporter) [reply typescript/infra/src/tx/transaction-reader.ts]: yeah just moved to something type safe on the interface

File: typescript/infra/scripts/safes/ethereum-safe-tx-oct-30.json
Comment (paulbalaji) [inline typescript/infra/scripts/safes/ethereum-safe-tx-oct-30.json:1]: delet

File: .changeset/new-olives-applaud.md
Comment (paulbalaji) [inline .changeset/new-olives-applaud.md:1]: think this is from a different PR

File: typescript/infra/src/tx/transaction-reader.ts [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+
+import {
+  icaOwnerChain,
+  icas,
+  safes,
+} from '../../config/environments/mainnet3/owners.js';
+import { DeployEnvironment } from '../config/environment.js';
+import { getSafeAndService } from '../utils/safe.js';
+
+export class TransactionReader {
```
Comment (paulbalaji) [inline typescript/infra/src/tx/transaction-reader.ts:38]: probably a nit but feel like the functionality here is more of a `SafeTxChecker` than just a reader, wdyt?

**Review by paulbalaji — COMMENTED** [review]
logic mostly lgtm, think there's places we can split up the class for easier reading but we can cross that bridge when we firm up the functionality here in the future

File: typescript/infra/src/tx/transaction-reader.ts [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+      return this.multiSendCallOnlyAddressCache[chain];
+    }
+
+    const safe = safes[chain];
+    if (!safe) {
+      return undefined;
+    }
+
+    const { safeSdk } = await retryAsync(() =>
+      getSafeAndService(chain, this.multiProvider, safe),
```
Comment (paulbalaji) [inline typescript/infra/src/tx/transaction-reader.ts:450]: can move the retryAsync logic into `getSafeAndService` directly - think it's only required when doing the `getSafe` call?
```ts
export async function getSafeAndService(
  chain: ChainNameOrId,
  multiProvider: MultiProvider,
  safeAddress: Address,
) {
  const safeSdk: Safe.default = await retryAsync(
    () => getSafe(chain, multiProvider, safeAddress),
    5,
    1000,
  );
  const safeService: SafeApiKit.default = getSafeService(chain, multiProvider);
  return { safeSdk, safeService };
}
```

File: typescript/infra/src/tx/transaction-reader.ts (hunk 2 above)
Comment (tkporter) [reply typescript/infra/src/tx/transaction-reader.ts:38]: it's not overly Safe specific, like the fetching of txs from the Safe API lives outside this class. Maybe `GovernTransactionReader`?

**Review by paulbalaji — APPROVED** [review]
don't forget to update the example command in the description as well

**Pull request merged**